### PR TITLE
HLRC: Fix '+' Not Correctly Encoded in GET Req.

### LINKED ISSUE #33077: Elastic Search high level client not encoding + symbol properly in Get Request
**Elasticsearch version** (`bin/elasticsearch --version`): 6.3.0

**Plugins installed**: [Kibana, Cerebro]

**JVM version** (`java -version`): 1.8

**OS version** (`uname -a` if on a Unix-like system): Linux kernel 4.9.93-0.1 #1 SMP Thu Jun 21 13:50:19 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:

**Steps to reproduce**:
Use rest high level client and         

where document id is any document id which contains "+" symbol.
GetRequest getRequest = new GetRequest(index, type, documentId);

        GetResponse getResponse = restHighLevelClient.get(getRequest);

Please include a *minimal* but *complete* recreation of the problem, including
(e.g.) index creation, mappings, settings, query etc.  The easier you make for
us to reproduce it, the more likely that somebody will take the time to look at it.

 1. Create any index and any type 
 2. Index a document with document id that contains "+"
 3. Search with the above provided code
4. You should not find the document with an id that includes "+"

**Provide logs (if relevant)**:

### PR DESCRIPTION
**Pull request by original-brownbear**
* Encode `+` correctly as `%2B`
* Closes #33077

### CODE REVIEW

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java
Comment (jasontedor) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: Why are we doing this manually and piecemeal instead of using `URLEncoder`?

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left a question.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java
Comment (original-brownbear) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: @javanna yea now that you mention it, I think the problem is with the code that handles the request, not the request itself.
`+` doesn't need to be encoded in a path component.

=> this PR and the issue seem to be missing the actual problem?

Comment (jasontedor) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: A `+` does not need to be encoded by the specifications indeed. However, the issue here is that we treat `+` specially, converting it to a ` `. This is because some browsers encode spaces as `+`. See: https://github.com/elastic/elasticsearch/blob/f7a9186372edcd9da31a89ac0565d53874db1839/server/src/main/java/org/elasticsearch/rest/RestUtils.java#L166-L168

Comment (jasontedor) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: I think the actual problem is that if you have an ID that contains a `+` and you try to use the HLRC to get a document by that then since we convert `+` to space server side then you can not get that document by ID unless you also encode the `+`.

Comment (original-brownbear) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: @jasontedor yes exactly that is the issue. So do we do the encoding here to account for the non-standard server behaviour?

Comment (javanna) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: interesting, is this something that we want to fix/change on the server-side rather than adapting the client then?

Comment (jasontedor) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: Yes, I think it is worth us making that change although it's a breaking one so we have to be careful and age it in  slowly, and maybe provide a system property for BWC for a time period.

Comment (jasontedor) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: Section 2.4 of RFC 3986 is [germane](https://tools.ietf.org/html/rfc3986#section-2.4) here. Since we never treat `+` as a sub-component delimiter (unlike, e.g., `&`) we do not need to require that it be encoded when it is to be treated as data. Treating a plus as a space is only required by the specification in the query string component of `application/x-www-form-urlencoded` requests. Since we do not handle this media type for any requests with bodies anyway, we never need to treat `+` as ` `. Our handling here is just odd, I think it's safe for us to remove this legacy behavior.

Comment (hub-cap) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: Im +1 to @jasontedor logic here.

**Review by jasontedor — CHANGES_REQUESTED** [review]
I think that we should start with a deprecation in 6.x, and probably a system property introduced there to cut over to the new behavior immediately. Then we can work to remove this in master. We also need to let the reporter on the original issue know how to workaround for today. Finally, we should consider opening an issue with this as our plan for feedback from the community in case there is a compelling reason that we are collectively missing for making this change. An issue will give it more visibility than a PR discussion where the title does not really reveal what we are planning to do here now.

**Comment by original-brownbear** [comment]
@jasontedor Looks like there already is an issue for this: https://github.com/elastic/elasticsearch/issues/5341

**Comment by casper1149** [comment]
I also noticed that IndexRequest incorrectly handles the "+" symbol, is it going to be fixed as a part of this PR or there is a separate one for the IndexRequest?

**Comment by Selikoff** [comment]
Can someone decide on a fix and merge the code?  I'm experiencing this exact issue where if I sent a + it is converted to space in the key and breaking linkage between servers.  I'm not sure what the correct solution is, but it's definitely broken as is.

**Comment by original-brownbear** [comment]
Jenkins test this

**Comment by original-brownbear** [comment]
@jasontedor @hub-cap sorry for the long pause here, brought this one back to life now :)
I added a system property to make the new behaviour optional but ran into a bit of a tricky spot with the SAML auth code. Maybe you guys can take a look and/or know more about what the correct approach is there? :)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutRequestHandler.java
Comment (hub-cap) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutRequestHandler.java]: I read some brief SO/docs on RelayState and it seems like we should not be doing decoding/encoding on it, but instead storing it as an "opaque object"... I think that it might be a correct assumption to not use the `decodeQueryString` here, but I dont know if we should take my word as gospel. Ill let @jasontedor comment further.

**Comment by casper1149** [comment]
@jloyolask8 I am not ES developer, so I can’t tell about the solution, but if you need a fix asap you can use the low level rest client as a workaround until this issue in the high level rest client is resolved.

**Comment by original-brownbear** [comment]
@jasontedor can you take another look here, specifically at https://github.com/elastic/elasticsearch/pull/33164#discussion_r236353154 ? Thanks!

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java
Comment (jloyolask8) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: > "what needs to be encoded should already be encoded and that is why we create the URI instance to encode each part."

That cannot be accurate. If I encode the data before calling this, is going to be re-encoded.
So this code should be the responsible for encoding and it should happen only once. 

Why are you using URI ? and Not URLEncoder?

**Comment by original-brownbear** [comment]
@javanna revived this one :)

**Comment by javanna** [comment]
heya @original-brownbear thanks for reviving this one. I'd prefer if we find somebody from the core team to review this one, given that it involves a breaking change, sysprop etc.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutRequestHandler.java
Comment (jasontedor) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutRequestHandler.java]: From what I have read, URL encoding is correct here, using `RestUtils` to decode looks wrong, and so the change here is valid. Let us validate this with @tvernum though.

File: server/src/main/java/org/elasticsearch/rest/RestUtils.java
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/rest/RestUtils.java]: I think that we should immediately deprecate a true value here. Then in 8.0.0 we can reject true, allow for unset or false (with the latter deprecated), and then finally remove in 9.0.0 (let’s add assertions on the major version for this). Let us also add a note to the migration guide. That means after this PR is merged to master and backported to 7.x (only), we need a follow-up in master to reject false and do the deprecation and add the assertion.

**Review by jasontedor — APPROVED** [review]
@original-brownbear I am sorry for the long delay on this one, it fell off my radar and I’ll aim to to better. This change looks good, thanks for carrying the torch on it. I left one comment, assuming that you add the deprecation and migration note, no need for another round on my end. I won’t be around for a few weeks to review a follow-up, no need to block on me for that, so feel free to ask for another reviewer from core/infra.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutRequestHandler.java
Comment (tvernum) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutRequestHandler.java]: I don't think is correct.
This string is already (mostly) decoded in `parseQueryStringAndValidateSignature` above, so any %-encoded entities will be expanded at that point.
If we `decode` here we will be double decoding.

Specifically the RelayState of `99%3a` would be encoded on the wire as `RelayState=99%253a`.
Then `parseQueryStringAndValidateSignature` will parse that back to `99%3a`
If we then try to `UrlDecoder.decode` that, it would end up at `99:` which is incorrect.

I'll need to look into what we ought to be doing with `+`, but I don't think this change is correct.

Comment (original-brownbear) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlLogoutRequestHandler.java]: Thanks @tvernum I think I figured this out actually and was able to revert this change.

The problem actually came from the assumption that we made here https://github.com/elastic/elasticsearch/pull/33164#discussion_r212990155 (never handling form data). It turns out that the form data encoding logic (`+` -> ` `) applies to parameters in the URL. I now made the changes to the logic to decode `+` to a space in query params only and this test passes again without any changes to the code here :)

**Comment by original-brownbear** [comment]
@jasontedor no worries, there is actually a silver lining to this delay :)
See here: https://github.com/elastic/elasticsearch/pull/33164#discussion_r296611521

This wasn't visible in our tests when I last worked on this, but now we have a few new REST tests that actually run search in line queries with spaces in them and those failed with the changes here. Adjusting the logic to properly encode/decode spaces as `+` and vice versa in query parameters fixed this. We can get back to it when you're back I guess :)

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by original-brownbear** [comment]
Thanks @jasontedor !

**Comment by falu2010-netflix** [comment]
Is there a fix rolled out in older version of client 5.6.x? If not how can I deal with it?

**Comment by original-brownbear** [comment]
@falu2010-netflix I'm afraid this fix will not be back-ported to older versions.
As a workaround, encoding `+` signs in doc ids as `%2B` is you best option I think.

**Pull request merged**
